server/src/main/java/org/elasticsearch/search/dfs/DfsSearchResult.java

```diff
@@ -135,24 +135,24 @@ public void writeTo(StreamOutput out) throws IOException {
         writeTermStats(out, termStatistics);
         writeFieldStats(out, fieldStatistics);
         out.writeVInt(maxDoc);
-        if (out.getVersion().onOrAfter(Version.V_7_10_0)) {
+        if (out.getTransportVersion().onOrAfter(TransportVersion.V_7_10_0)) {
             out.writeOptionalWriteable(getShardSearchRequest());
         }
-        if (out.getVersion().onOrAfter(Version.V_8_4_0)) {
-            if (out.getVersion().onOrAfter(Version.V_8_7_0)) {
+        if (out.getTransportVersion().onOrAfter(TransportVersion.V_8_4_0)) {
+            if (out.getTransportVersion().onOrAfter(TransportVersion.V_8_7_0)) {
                 out.writeOptionalCollection(knnResults);
             } else {
                 if (knnResults != null && knnResults.size() > 1) {
                     throw new IllegalArgumentException(
-                        "Versions before 8.7.0 don't support multiple [knn] search clauses and search was sent to ["
-                            + out.getVersion()
+                        "Versions before 8070099 don't support multiple [knn] search clauses and search was sent to ["
```

Comment:
Do we want to keep `Version` here?

Reply:
I don't think we want Version, it wouldn't make sense. However, this exception seems odd. It looks like some validation that should have happened before we are trying to serialize results. Perhaps someone from the Search area can clarify why such validation is happening so late.

Reply:
@benwtrent What is the exception here needed for?

Reply:
This is the exact same validation that occurs up in `SearchSourceBuilder`. We cannot allow serializing this to older nodes if there are more than one KNN search results. Similarly, we don't allow search source builder to serialize to older nodes if it has more than one KNN search clause.

I fail to see how this is a problem here, but NOT a problem in the SearchSourceBuilder.

Reply:
It's not a problem, it's more a question on the importance of the message containing the actual self managed node version, vs an opaque version for the transport protocol. What I find odd here is this error being thrown as an IllegalArgumentException, which will go back to users as a 400 http response. Normally such errors should be checked long before we are serializing results.

Reply:
I am happy to make it NOT an illegal argument exception and to remove the node version. We can make it say "Cannot serialize multiple KNN results to node with version: " and use the streamable version.

But we should keep this check here. I would rather keep it as doing this would be VERY wrong.

Reply:
The "to node with version:" is the problem. This won't make sense in serverless. Asking my question in a different way: is it important this is actually the version of the code running, or simply diagnostic for an ES dev looking at the message (in which case transport version should be ok)?

Reply:
We can be handwavy and simply say "older node" or whatever. But for debugging, it would be good to know the version that this capability was added. Otherwise, somebody debugging a failure like this would have to spend time figuring out the feature release and find a what versions they care about.

I hope we don't remove debuggability for the sake of flexibility.

Reply:
If this is only a stopgap exception aimed at developers only, then we can leave it as the transport version

Reply:
> If this is only a stopgap exception aimed at developers only, then we can leave it as the transport version

Non-administrators wouldn't care about this exception, exactly. Administrators or folks debugging failures will care.

Reply:
I think a problem of identifying a "code version running" is not only limited to this one exception here. We will have to have some kind of system with a matrix of transport version - commit/onprem version. 

Perhaps for this case specifically we change this as per @benwtrent suggestion. "Cannot serialize multiple KNN results to older node with transport version:" or similar (the 'older node' feels strange)

Reply:
I've created #93387 to check exception messages now using transport version

Reply:
I've updated the exception message